process/smartContract/scQueryService.go

```diff
@@ -116,60 +176,204 @@ func (service *SCQueryService) shouldAllowQueriesExecution() bool {
+		if err != nil {
+			return nil, nil, err
+		}
+
+		header := service.getBlockHeader(blockHeader)
+		return header, header.GetRootHash(), nil
+	}
+
+	return service.mainBlockChain.GetCurrentBlockHeader(), service.mainBlockChain.GetCurrentBlockRootHash(), nil
+}
+
+func (service *SCQueryService) getBlockHeaderByHash(headerHash []byte) (data.HeaderHandler, error) {
+	epoch, err := service.getOptionalEpochByHash(headerHash)
+	if err != nil {
+		return nil, err
+	}
+
+	header, err := service.getBlockHeaderInEpochByHash(headerHash, epoch)
+	if err != nil {
+		return nil, err
+	}
+
+	return header, nil
+}
+
+func (service *SCQueryService) getOptionalEpochByHash(hash []byte) (core.OptionalUint32, error) {
+	if !service.historyRepository.IsEnabled() {
+		return core.OptionalUint32{}, nil
+	}
+
+	epoch, err := service.historyRepository.GetEpochByHash(hash)
+	if err != nil {
+		return core.OptionalUint32{}, err
+	}
+
+	return core.OptionalUint32{Value: epoch, HasValue: true}, nil
+}
+
+func (service *SCQueryService) getBlockHeaderInEpochByHash(headerHash []byte, epoch core.OptionalUint32) (data.HeaderHandler, error) {
+	shardId := service.shardCoordinator.SelfId()
+	unitType := dataRetriever.GetHeadersDataUnit(shardId)
+	storer, err := service.storageService.GetStorer(unitType)
+	if err != nil {
+		return nil, err
+	}
+
+	var headerBuffer []byte
+
+	if epoch.HasValue {
+		headerBuffer, err = storer.GetFromEpoch(headerHash, epoch.Value)
+	} else {
+		headerBuffer, err = storer.Get(headerHash)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	header, err := process.UnmarshalHeader(shardId, service.marshaller, headerBuffer)
+	if err != nil {
+		return nil, err
+	}
+
+	return header, nil
+}
+
+func (service *SCQueryService) getBlockHeader(currentHeader data.HeaderHandler) data.HeaderHandler {
+	header, _, err := service.getBlockHeaderByNonce(currentHeader.GetNonce() + 1)
+	if err != nil {
+		return currentHeader
+	}
+
+	return header
+}
+
+func (service *SCQueryService) getBlockHeaderByNonce(nonce uint64) (data.HeaderHandler, []byte, error) {
+	headerHash, err := service.getBlockHashByNonce(nonce)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	header, err := service.getBlockHeaderByHash(headerHash)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return header, headerHash, nil
+}
 
-	return vmOutput, nil
+func (service *SCQueryService) getBlockHashByNonce(nonce uint64) ([]byte, error) {
+	shardId := service.shardCoordinator.SelfId()
+	hashByNonceUnit := dataRetriever.GetHdrNonceHashDataUnit(shardId)
+
+	return process.GetHeaderHashFromStorageWithNonce(
+		nonce,
+		service.storageService,
+		service.uint64ByteSliceConverter,
+		service.marshaller,
+		hashByNonceUnit,
+	)
 }
 
 func (service *SCQueryService) checkForRootHashChanges(rootHashBefore []byte) error {
-	rootHashAfter := service.blockChain.GetCurrentBlockRootHash()
+	rootHashAfter := service.mainBlockChain.GetCurrentBlockRootHash()
```

Comment:
shouldn't it verify this against the api blockchain block roothash?

I would only use the mainBlockChain to set the apiBlockChain data when the API is called without block parameters (so for current)